### PR TITLE
Suspend game thread after it has been running in the background for a…

### PR DESCRIPTION
**Pull request by bofeng-song**

### CODE REVIEW

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 1]
```diff
@@ -376,13 +382,26 @@ class GameInputProxy {
...
+    void registerAppEventCallback(AppEventCallback callback) {
+        _eventCallback = std::move(callback);
     }
 
     bool isAnimating() const {
         return _isVisible && _hasWindow;
     }
 
+    bool isLaunched() const {
+        return _launched;
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp]: Rename `_launched` to `_isInited`? I think `_isInited` is better.

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 2]
```diff
@@ -483,12 +514,18 @@ int32_t AndroidPlatform::loop() {
...
-        flushTasksOnGameThreadJNI();
-        if (_inputProxy->isAnimating()) {
+        if (_inputProxy->isAnimating() ) {
+            runTask();
             flushTasksOnGameThreadAtForegroundJNI();
         }
+        flushTasksOnGameThreadJNI();
+        if (_isLowFrequencyLoopEnabled) {
+            //Suspend a game thread after it has been running in the background for a specified amount of time
+            if (_lowFrequencyTimer.getSeconds() > LOW_FREQUENCY_EXPIRED_DURATION_SECONDS) {
```
Comment (PatriceJiang) [inline native/cocos/platform/android/AndroidPlatform.cpp:524]: 切后台一段时间挂起, 不执行任何逻辑会有一些副作用.  比如用户网络连接有心跳逻辑, 挂起后不在执行, 会导致断联. 一些基于位置信息的应用 也会切换到后台, 出现无法上报位置信息. 这个配置默认关闭, 由开发者开启是否更合理? @minggo @bofeng-song

File: native/cocos/platform/android/AndroidPlatform.cpp [hunk 3]
```diff
@@ -430,6 +449,18 @@ AndroidPlatform::~AndroidPlatform() = default;
 int AndroidPlatform::init() {
     cc::FileUtilsAndroid::setassetmanager(_app->activity->assetManager);
     _inputProxy = ccnew GameInputProxy(this);
+    _inputProxy->registerAppEventCallback([this](int32_t cmd) {
+        if (APP_CMD_START == cmd || APP_CMD_INIT_WINDOW == cmd) {
+            if (_inputProxy->isAnimating()) {
+                _isLowFrequencyLoopEnabled = false;
+                _loopTimeOut = 0;
+            }
+        } else if(APP_CMD_STOP == cmd) {
```
Comment (minggo) [inline native/cocos/platform/android/AndroidPlatform.cpp:456]: Only enter low frequency mode when stop? When `_inputProxy->isAnimating()` is false, then should enter frequency mode?

File: native/cocos/platform/android/AndroidPlatform.cpp (hunk 2 above)
Comment (minggo) [reply native/cocos/platform/android/AndroidPlatform.cpp:524]: @PatriceJiang please use English in future.

File: native/cocos/platform/android/AndroidPlatform.cpp (hunk 3 above)
Comment (bofeng-song) [reply native/cocos/platform/android/AndroidPlatform.cpp:456]: When the total low frequency duration is exhausted, the game thread will be suspended

File: native/cocos/platform/android/AndroidPlatform.cpp (hunk 2 above)
Comment (bofeng-song) [reply native/cocos/platform/android/AndroidPlatform.cpp:524]: Applications running in the background also need heartbeat packets, this feature should belong to a small number of user requirements. So we can turn on by default.

**Review by PatriceJiang — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Pull request merged**